src/OrchardCore.Modules/OrchardCore.Media/Services/MediaFileStoreResolverMiddleware.cs

```diff
@@ -0,0 +1,180 @@
+            {
+                await _next(context);
+                return;
+            }
+
+            var validatePath = context.Request.Path.StartsWithSegments(_assetsRequestPath, StringComparison.OrdinalIgnoreCase, out var subPath);
+            if (!validatePath)
+            {
+                _logger.LogDebug("Request path {Path} does not match the assets request path {RequestPath}", subPath, _assetsRequestPath);
+                await _next(context);
+                return;
+            }
+
+            // TODO optional.
+            // This will not cache the file if the file extension is not supported.
+            var fileExtension = GetExtension(subPath);
+            if (!_allowedFileExtensions.Contains(fileExtension, StringComparer.OrdinalIgnoreCase))
+            {
+                _logger.LogDebug("File extension not supported for request path {Path}", subPath);
+                await _next(context);
+                return;
+            }
+
+
+            var fileInfo = _mediaFileStoreCacheProvider.GetFileInfo(subPath);
+            if (fileInfo.Exists)
+            {
+                // TODO Consider how to provide a graceful cache period on this, 
+                // so we do not serve a file from cache, if it has been deleted from Blob storage.
+                // One idea: Treat this cache like any other Cdn cache. Cache bust it.
+                // We would need another memory cache here of file hashes for cached files.
+                // We could then compare the cached file hash with the file hash
+                // from the IFileStoreVersionProvider to see if the hash is the same.
+                // If it is not, then remove the file from the disc cache.
+                // The IFileStoreVersionProvider then becomes responsible for
+                // gracefully expiring it's own memory cache of version hashes.
+                // However heavy on memory cache, for a large quantity of files, and requires much 
+                // initial creation of sha hashes.
+
+                // Date comparison is also an option, with the IFileStoreVersionProvider also caching
+                // last modified dates, and matching here. Not sure the dates will match however.
+
+                // More thinking to be done.
+
+                // When multiple requests occur for the same file the download 
+                // may already be in progress so we wait for it to complete.
+                if (_writeTasks.TryGetValue(subPath, out var writeTask))
+                {
+                    await writeTask.Value;
+                }
+
+                await _next(context);
+                return;
+            }
+
+            // When multiple requests occure for the same file we use a Lazy<Task>
+            // to initialize the file store request once.
+            await _writeTasks.GetOrAdd(subPath, x => new Lazy<Task>(async () =>
```

Comment:
This logic might still be flowed in case the task throws an exception. I assume the worst case is that you get a 404 and the request can be resent, but there are ways to fix it. Don't worry about it for now, I'll show you at some point.

Reply:
Yes, at the moment it just traps exceptions, and logs them, assuming a 404 later in the pipeline.